### PR TITLE
Changed id-writer to save join table records based on association pri…

### PR DESCRIPTION
**Pull request by himesh-r**
…mary key #20995

### CODE REVIEW

File: activerecord/lib/active_record/associations/collection_association.rb [hunk 1]
```diff
@@ -72,7 +72,8 @@ def ids_writer(ids)
         pk_type = reflection.primary_key_type
         ids = Array(ids).reject(&:blank?)
         ids.map! { |i| pk_type.cast(i) }
-        replace(klass.find(ids).index_by(&:id).values_at(*ids))
+        records = klass.where(reflection.association_primary_key => ids).to_a.index_by(&reflection.association_primary_key.to_sym).values_at(*ids)
```
Comment (sgrif) [inline activerecord/lib/active_record/associations/collection_association.rb]: We shouldn't need to call `.to_a` here.

Comment (sgrif) [inline activerecord/lib/active_record/associations/collection_association.rb]: Can you break this into multiple lines?

**Comment by sgrif** [comment]
A few nitpicks. Looks good otherwise. Can you update, and also squash your commits?

File: activerecord/lib/active_record/associations/collection_association.rb (hunk 1 above)
Comment (himesh-r) [inline activerecord/lib/active_record/associations/collection_association.rb]: Done, check the new commit please.

**Comment by himesh-r** [comment]
I have made changed and squashed the commits. Please let me know if anything else is required.

**Comment by maclover7** [comment]
Are we ready to roll here, @sgrif?

**Comment by arthurnn** [comment]
merged via https://github.com/rails/rails/commit/bcd0c8cfc250039ddc145c3302633c4890880930 
thanks

**Comment by arthurnn** [comment]
changelog added https://github.com/rails/rails/commit/d8a53b49a7490eb0115fcf6f7f4c872a5393e251

**Pull request closed without merging**